File: Array/200-number-of-islands/number-of-islands.java
class Solution {

    private int ROWS;
    private int COLS;
    private Set<String> visited;
    private char[][] grid;

    public void dfs(int r, int c) {
        visited.add(r+"-"+c);
        int[][] dirs = {{0,-1}, {0, 1}, {-1,0}, {1,0}};

        for(int[] dir: dirs) {
            int new_r = r + dir[0];
            int new_c = c + dir[1];

            if(new_r>=0 && new_r<ROWS && new_c>=0 && new_c<COLS && grid[new_r][new_c] == '1' && !visited.contains(new_r + "-" + new_c)) {
                dfs(new_r, new_c);
            }

        }
    }

    public int numIslands(char[][] grid) {
        ROWS = grid.length;
        COLS = grid[0].length;
        visited = new HashSet<>();
        this.grid = grid;
        int islandsCount = 0;

        for(int i=0;i<ROWS; i++) {
            for(int j=0;j<COLS;j++) {
                if(grid[i][j] == '1' && !visited.contains(i+"-" + j)) {
                    dfs(i,j);
                    islandsCount++;
                }
            }
        }
        return islandsCount;
        


    }
}

File: Array/200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])

        islands = 0
        visited = set()
        def dfs(r, c):
            visited.add((r,c))

            dir = [(-1, 0), (1,0), (0,-1), (0,1)]

            for i, j in dir:
                new_r = i + r
                new_c = j + c
                if new_r in range(ROWS) and new_c in range(COLS) and grid[new_r][new_c] == "1" and (new_r, new_c) not in visited:
                    dfs(new_r, new_c)

        
        for i in range(ROWS):
            for j in range(COLS):
                if grid[i][j] == "1" and (i,j) not in visited:
                    islands+=1
                    dfs(i,j)

        return islands
